src/wildcard.rs

```diff
@@ -382,10 +383,36 @@ fn wildcard_test_flags_then_complete(
         .check_type()
         .map(|x| x == DirEntryType::reg)
         .unwrap_or(false);
-    if executables_only && (!is_regular_file || waccess(filepath, X_OK) != 0) {
+    let is_executable = is_regular_file && waccess(filepath, X_OK) == 0;
+    if executables_only && !is_executable {
         return false;
     }
 
+    // For executables on Cygwin, prefer the name without the .exe, to match
+    // better with Unix names, but only if there isn't also a file without that
+    // extension and the user hasn't started to type the extension
+    let filename = if cfg!(cygwin)
+        && is_executable
+        && string_suffixes_string_case_insensitive(L!(".exe"), filename)
+        && wc.len() + 4 <= filename.len()
+    {
+        let filepath_stripped = &filepath[0..filepath.len() - 4];
+        let stat_stripped = lwstat(filepath_stripped).map(|stat| (stat.dev(), stat.ino()));
+        let stat = lwstat(filepath).map(|stat| (stat.dev(), stat.ino()));
```

Comment:
I guess we could cache this lwstat ourselves since
we might call it again,
so maybe we should have a `let file_metadata = Lazy::new(|| lwstat(filepath).ok());` ?